README.md
# NAME

Device::LPS331AP - I2C interface to LPS331AP Thermometer and Barometer using Device::SMBus

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shantanubhadoria/perl-Device-LPS331AP"><img src="https://api.travis-ci.org/shantanubhadoria/perl-Device-LPS331AP.svg?branch=build/master" alt="Travis status" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Device-LPS331AP%200.006"><img src="https://badgedepot.code301.com/badge/cpantesters/Device-LPS331AP/0.006" alt="CPAN Testers result" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Device-LPS331AP-0.006"><img src="https://badgedepot.code301.com/badge/kwalitee/Device-LPS331AP/0.006" alt="Distribution kwalitee" /></a>
    <a href="https://gratipay.com/shantanubhadoria"><img src="https://img.shields.io/gratipay/shantanubhadoria.svg" alt="Gratipay" /></a>
    </p>
</div>

# VERSION

version 0.006

# ATTRIBUTES

## I2CBusDevicePath

this is the device file path for your I2CBus that the LPS331AP is connected on e.g. /dev/i2c-1
This must be provided during object creation.

## Altimeter

    $self->Altimeter->enable();
    $self->Altimeter->getReading();

This is a object of [Device::Altimeter::LPS331AP](https://metacpan.org/pod/Device::Altimeter::LPS331AP)

# SUPPORT

## Bugs / Feature Requests

Please report any bugs or feature requests through github at 
[https://github.com/shantanubhadoria/perl-device-lps331ap/issues](https://github.com/shantanubhadoria/perl-device-lps331ap/issues).
You will be notified automatically of any progress on your issue.

## Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

[https://github.com/shantanubhadoria/perl-device-lps331ap](https://github.com/shantanubhadoria/perl-device-lps331ap)

    git clone git://github.com/shantanubhadoria/perl-device-lps331ap.git

# AUTHOR

Shantanu Bhadoria &lt;shantanu at cpan dott org>

# CONTRIBUTOR

Shantanu <shantanu@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Changelog for Device-LPS331AP

0.006 2016-04-07T06:42:36
 - No changes found

0.005 2015-03-25T07:29:34
 - v0.005 (Shantanu Bhadoria)
 - fixed extends removed commented code, why the heck did I do that??
   (Shantanu Bhadoria)

0.004 2013-09-27T10:29:46
 - v0.004 (Shantanu)
 - Resolved xt test failures (Shantanu)
 - Refactoring Submodule to allow easy loading of similar subdevices as
   plugins using a common prefix in MOP (Shantanu Bhadoria)
 - Minor Changes (Shantanu Bhadoria)

0.003 2013-09-12T09:22:34
 - v0.003 (Shantanu)
 - Added README.pod and Changes (Shantanu)

0.002 2013-09-11T09:27:04
 - pod fix (Shantanu)

0.001 2013-09-11T09:08:21
 - first push (Shantanu)
 - First push (Shantanu Bhadoria)
 - first fixes (Shantanu Bhadoria)
 - First few working changes (Shantanu Bhadoria)
 - first commit (Shantanu Bhadoria)


lib/Device/LPS331AP/Altimeter.pm
package Device::LPS331AP::Altimeter;

# PODNAME: Device::LPS331AP::Altimeter 
# ABSTRACT: I2C interface to LPS331AP Altimeter using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

# Registers for the Altimeter 
use constant {
    CTRL_REG1 => 0x20,
};

# X, Y and Z Axis magnetic Field Data value in 2's complement
use constant {
    PRESS_OUT_H  => 0x2a,
    PRESS_OUT_L  => 0x29,
    PRESS_OUT_XL => 0x28,

    TEMP_OUT_H => 0x2c,
    TEMP_OUT_L => 0x2b,
};

#use integer; # Use arithmetic right shift instead of unsigned binary right shift with >> 4

extends 'Device::SMBus';

=attr I2CDeviceAddress

Containd the I2CDevice Address for the bus on which your altimeter is connected. It would look like 0x6b. Default is 0x5d.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x5d,
);

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

CTRL_REG1
power: {0:off,1:on}
ODR2:output data rate
ODR1
ODR0
DIFF_EN: Interrupt generation{0:disabled,1:enabled}
BDU: Block Data Update{0:continuous update, 1: output registers not updated until MSB and LSB reading}
DELTA_EN: (1: delta pressure registers enabled. 0: disable)
SIM: SPI Serial Interface Mode selection (0: 4-wire interface; 1: 3-wire interface)
default = 1 111 0 1 1 0
=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1,0b11100000);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=cut

sub getRawReading {
    my ($self) = @_;

    return (
        pressure => $self->readNBytes(PRESS_OUT_XL,3),
        temperature => ( $self->_typecast_int_to_int16( ($self->readByteData(TEMP_OUT_H) << 8) | $self->readByteData(TEMP_OUT_L) ) ) ,
        temp => $self->readNBytes(TEMP_OUT_L,2) ,
    );
}

sub getPressureMillibars{
    my ($self) = @_;
    return $self->readNBytes(PRESS_OUT_XL,3)/4096;
}

sub getPressureInchesHg{
    my ($self) = @_;
    return $self->readNBytes(PRESS_OUT_XL,3)/138706.5;
}

=head2

converts pressure in mbar to altitude in meters, using 1976 US
Standard Atmosphere model (note that this formula only applies to a
height of 11 km, or about 36000 ft)
If altimeter setting (QNH, barometric pressure adjusted to sea
level) is given, this function returns an indicated altitude
compensated for actual regional pressure; otherwise, it returns
the pressure altitude above the standard pressure level of 1013.25
mbar or 29.9213 inHg

QNH is the Barometric pressure adjusted to sea level for a particular region. This value helps altitude corrections based on base barometric pressure in your region.

(a-((x/b)^c))*d;
d - (x^c)*(d/b^c);
=cut

sub getPressureToAltitudeMeters {
    my ($self,$pressure,$qnh) = @_;
    $qnh |= 1013.25;
    my $altitude = (1-(($pressure/$qnh)**(0.190263)))*44330.8;
}

sub getTemperatureCelsius{
    my ($self) = @_;
    
    return 42.5 +  $self->_typecast_int_to_int16($self->readNBytes(TEMP_OUT_L,2))/480;
}

sub getTemperatureFarenheit{
    my ($self) = @_;
    
    return 108.5 +  $self->_typecast_int_to_int16($self->readNBytes(TEMP_OUT_L,2))/480 * 1.8;
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

sub _typecast_int_to_int32 {
    return  unpack 'ss' => pack 'SS' => $_[1];
}

sub calibrate {
    my ($self) =@_;

}

1;


lib/Device/LPS331AP/Thermometer.pm
package Device::LPS331AP::Altimeter;

# PODNAME: Device::LPS331AP::Thermometer 
# ABSTRACT: I2C interface to LPS331AP Thermometer using Device::SMBus
# COPYRIGHT
# VERSION

use 5.010;
use Moose;
use POSIX

# Registers for the Accelerometer 
use constant {
    CTRL_REG1 => 0x20,
};

# X, Y and Z Axis magnetic Field Data value in 2's complement
use constant {
    PRESS_OUT_H => 0x2a,
    PRESS_OUT_L => 0x29,

    TEMP_OUT_H => 0x2c,
    TEMP_OUT_L => 0x2b,
};

use integer; # Use arithmetic right shift instead of unsigned binary right shift with >> 4

extends 'Device::SMBus';

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x5d,
);

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

CTRL_REG1
power: {0:off,1:on}
ODR2:output data rate
ODR1
ODR0
DIFF_EN: Interrupt generation{0:disabled,1:enabled}
BDU: Block Data Update{0:continuous update, 1: output registers not updated until MSB and LSB reading}
DELTA_EN: (1: delta pressure registers enabled. 0: disable)
SIM: SPI Serial Interface Mode selection (0: 4-wire interface; 1: 3-wire interface)
default = 1 111 0 1 1 0
=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1,0b11110110);
    $self->writeByteData(CTRL_REG1,0b11110110);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=cut

sub getRawReading {
    my ($self) = @_;

    return (
        pressure    => ( $self->_typecast_int_to_int16( ($self->readByteData(PRESS_OUT_H) << 8) | $self->readByteData(PRESS_OUT_L) ) ) ,
        temperature => ( $self->_typecast_int_to_int16( ($self->readByteData(TEMP_OUT_H) << 8) | $self->readByteData(TEMP_OUT_L) ) ) ,
    );
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

sub calibrate {
    my ($self) =@_;

}

1;


lib/Device/Altimeter/LPS331AP.pm
use strict;
use warnings;
package Device::Altimeter::LPS331AP;

# PODNAME: Device::Altimeter::LPS331AP 
# ABSTRACT: I2C interface to LPS331AP Altimeter using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use POSIX;

use Moose;
extends 'Device::SMBus';

# Registers for the Altimeter

=register CTRL_REG1

=cut

use constant {
    CTRL_REG1 => 0x20,
};

# X, Y and Z Axis magnetic Field Data value in 2's complement

=register PRESS_OUT_H

=register PRESS_OUT_L

=register PRESS_OUT_XL

=register TEMP_OUT_H

=register TEMP_OUT_L

=cut

use constant {
    PRESS_OUT_H  => 0x2a,
    PRESS_OUT_L  => 0x29,
    PRESS_OUT_XL => 0x28,

    TEMP_OUT_H => 0x2c,
    TEMP_OUT_L => 0x2b,
};

=attr I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your altimeter is connected. It would look like 0x6b. Default is 0x5d.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x5d,
);

=method enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

CTRL_REG1
power: {0:off,1:on}
ODR2:output data rate
ODR1
ODR0
DIFF_EN: Interrupt generation{0:disabled,1:enabled}
BDU: Block Data Update{0:continuous update, 1: output registers not updated until MSB and LSB reading}
DELTA_EN: (1: delta pressure registers enabled. 0: disable)
SIM: SPI Serial Interface Mode selection (0: 4-wire interface; 1: 3-wire interface)
default = 1 111 0 1 1 0
=cut

sub enable {
    my ($self) = @_;
    $self->writeByteData(CTRL_REG1,0b11100000);
}

=method getRawReading

    $self->getRawReading()

Return raw readings from registers

=cut

sub getRawReading {
    my ($self) = @_;

    return (
        pressure => $self->readNBytes(PRESS_OUT_XL,3),
        temperature => ( $self->_typecast_int_to_int16( ($self->readByteData(TEMP_OUT_H) << 8) | $self->readByteData(TEMP_OUT_L) ) ) ,
        temp => $self->readNBytes(TEMP_OUT_L,2) ,
    );
}

=method getPressureMillibars

Get pressure in Millibars

=cut

sub getPressureMillibars{
    my ($self) = @_;
    return $self->readNBytes(PRESS_OUT_XL,3)/4096;
}

=method getPressureInchesHg

Get pressure in inches of mercury

=cut

sub getPressureInchesHg{
    my ($self) = @_;
    return $self->readNBytes(PRESS_OUT_XL,3)/138706.5;
}

=method getPressureToAltitudeMeters

converts pressure in mbar to altitude in meters, using 1976 US
Standard Atmosphere model (note that this formula only applies to a
height of 11 km, or about 36000 ft)
If altimeter setting (QNH, barometric pressure adjusted to sea
level) is given, this function returns an indicated altitude
compensated for actual regional pressure; otherwise, it returns
the pressure altitude above the standard pressure level of 1013.25
mbar or 29.9213 inHg

QNH is the Barometric pressure adjusted to sea level for a particular region. This value helps altitude corrections based on base barometric pressure in your region.

(a-((x/b)^c))*d;
d - (x^c)*(d/b^c);
=cut

sub getPressureToAltitudeMeters {
    my ($self,$pressure,$qnh) = @_;
    $qnh |= 1013.25;
    my $altitude = (1-(($pressure/$qnh)**(0.190263)))*44330.8;
}

=method getTemperatureCelsius

Get Temperature in degrees celsius

=cut

sub getTemperatureCelsius{
    my ($self) = @_;
    
    return 42.5 +  $self->_typecast_int_to_int16($self->readNBytes(TEMP_OUT_L,2))/480;
}

=method getTemperatureFahrenheit

Get Temperature in Fahrenheit

=cut

sub getTemperatureFahrenheit{
    my ($self) = @_;
    
    return 108.5 +  $self->_typecast_int_to_int16($self->readNBytes(TEMP_OUT_L,2))/480 * 1.8;
}

sub _typecast_int_to_int16 {
    return  unpack 's' => pack 'S' => $_[1];
}

sub _typecast_int_to_int32 {
    return  unpack 'ss' => pack 'SS' => $_[1];
}

=method calibrate

Placeholder for a calibration function

=cut

sub calibrate {
    my ($self) =@_;

}

1;
